### PR TITLE
AddParamTypeBasedOnPHPUnitDataProviderRector: Enhance existing rule to handle PHPUnit 10+ DataProvider Attribute

### PR DESCRIPTION
**Pull request by mcampbell508**
The purpose of this PR is to hopefully address https://github.com/rectorphp/rector/issues/8179.

I have tried to add atomic commits which should hopefully explain the changes made if you have time to review them commit by commit. I first started by adding failing tests as part of the first commit, which when running when checked out on that commit, fail.

I then refactored the existing code slightly to reuse the existing logic but with a renamed variable name, and then made the changes which I hope are along the right lines and within the expectations of the contributing guide.

The new method `getPhpDataProviderAttribute`, I added, is very similar to the existing code of https://github.com/rectorphp/rector-src/blob/77258018af10a7c23477831e0eab49dd7c0b2bfe/rules/Php80/NodeAnalyzer/PhpAttributeAnalyzer.php#L31-L44 but I thought instead of doing a `hasAttribute` check, I would rather just loop through the attributes and collect what was needed (If the Attribute is the one we want here).

I am also not sure if we have to add a check that the version of PHP should be at least 8.0 for this feature, I tried to do so but the existing configured rule logic caused my tests to fail when having the following check in the rule:

```
    private function getPhpDataProviderAttribute($node): ?Attribute
    {
+        if (!$this->phpVersionProvider->isAtLeastPhpVersion(PhpVersionFeature::ATTRIBUTES)) {
+            return null;
+        }
```

because of the existing code:

https://github.com/rectorphp/rector-src/blob/48febdb978201ffdd2d62ccb970ede09c3160134/rules-tests/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector/config/configured_rule.php#L12

I have left out the above `isAtLeastPhpVersion` check for now.

## Other notes

- Please see the commit and test I added for allowing only single data providers per test method. We may want to add this for the PHPDoc style of declaring data providers too, but I thought it may be out of scope for this PR, to do so. I hope you can see why this may be undesirable, because the types across different providers may be different - unless we did some sort of merging of all providers to determine the type.
- I also added a rather minor housekeeping update for the docs to do with a missing `;`, I can remove this if you would like it to be separate from this PR / want it gone.
- I am not sure if I was supposed to include the built docs as part of the PR, which I have done.

## Technical notes

- As mentioned in the PHPUnit 10 upgrade: https://github.com/sebastianbergmann/phpunit/issues/4502

> PHPUnit 10 will first look for metadata in attributes before it looks at comments

which we probably want to do here.

- An edge case that I may not have covered is for test methods that do the following (although it would be very unlikely but I guess could happen like anything :smile: ):

```
    /** @dataProvider provideData */
    #[\PHPUnit\Framework\Attributes\DataProvider('provideMoreData')]
    public function testAttributes($one, $two): void
    {
    }

    public static function provideData(): Iterator
    {
        yield ['string', 456];
    }

    public static function provideMoreData(): Iterator
    {
        yield [456, 'string'];
    }
```

Should we account for this?

Please let me know if you would like to make any changes, or please feel free to make any modifications to suit your needs/code style, if preferred.

I am fairly new to writing code in Rector but it is a rather enjoyable thing to work with.

### CODE REVIEW

File: rules-tests/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector/Fixture/fixture.php.inc [hunk 1]
```diff
@@ -38,6 +43,11 @@ final class SomeTestWithDataProvider extends TestCase
     {
     }
 
+    #[\PHPUnit\Framework\Attributes\DataProvider('provideData')]
+    public function test_with_attribute(string $name)
+    {
+    }
```
Comment (TomasVotruba) [inline rules-tests/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector/Fixture/fixture.php.inc]: Could you separate these changes to a new fixture file? It will be easier to connect together in years ahead blame

File: rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector.php [hunk 2]
```diff
@@ -85,6 +93,11 @@ public function test(string $value)
     {
     }
 
+    #[\PHPUnit\Framework\Attributes\DataProvider('provideData')]
+    public function test_with_attribute(string $name)
+    {
+    }
```
Comment (TomasVotruba) [inline rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector.php]: Feel free to keep only one example, so docs are short to read and point is passed. Feel free to pick one you prefere

File: rules-tests/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector/Fixture/fixture.php.inc (hunk 1 above)
Comment (TomasVotruba) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector/Fixture/fixture.php.inc]: Just friendly ping :)

Comment (mcampbell508) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector/Fixture/fixture.php.inc]: Hi - apologies, have just pushed up some changes.

Separated all tests into new files in the same directory as the original tests.

File: rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector.php (hunk 2 above)
Comment (mcampbell508) [reply rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector.php]: Thanks, removed the commit I added so it is the original code example for now

**Comment by mcampbell508** [comment]
@TomasVotruba please be aware one test started failing after the push before last:

https://github.com/rectorphp/rector-src/actions/runs/6282888032/job/17062803915#step:5:97

I changed the test `rules-tests/TypeDeclaration/Rector/ClassMethod/AddParamTypeBasedOnPHPUnitDataProviderRector/Fixture/skip_no_type_using_attribute.php.inc` to instead do something else to cause the unknown type. Please let me know if there is any issue with this or you would just like me to remove the test.

Change:

```diff
     public static function provideData(): array
     {
         return [
-            [ new class {} ],
+            [ self::SOMETHING ],
         ];
     }
 }

```

We cannot do similar as the existing test not using the attribute because public static providers in PHPUnit 10+ do not allow arguments passed, I believe.

**Comment by mcampbell508** [comment]
Hi @TomasVotruba - I have made some changes to accommodate for changes around this rule merged earlier today in https://github.com/rectorphp/rector-src/pull/5068. 

I had to refactor my original code slightly, but I think I have now accommodated for the changes @jlherren made. This includes adding 2 more tests around use of the attribute instead and also mixing between php doc and attribute.

The changes in https://github.com/rectorphp/rector-src/pull/5068 also address some concerns I added in the description of this merge request involving multiple providers, which is nice!

I believe I also made all other changes around your comments.

Have requested a re-review, thanks.

I am not sure on how you prefer to review in terms of marking comments as resolved, for now I have left that up to you.

**Comment by TomasVotruba** [comment]
Looks good, thank you 👍
